### PR TITLE
Versus mode

### PR DESCRIPTION
**Pull request by saralein**

### CODE REVIEW

File: lib/initializer.rb [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+def initializer(size, empty_char)
+  board = Board.new(size, empty_char)
+  checker = BoardChecker.new(size, empty_char)
+  script = Script.new
+  user_interface = UserInterface.new(board, script)
+
+  user_interface.welcome
+  mode_selection = user_interface.what_mode?
+
+  unless(mode_selection == 'computerVScomputer')
```
Comment (zdavy) [inline lib/initializer.rb]: have you considered migrating this string somewhere else?
(http://deviq.com/magic-strings/)

File: lib/initializer.rb [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+require_relative './board/board_checker.rb'
+require_relative './players/ai.rb'
+require_relative './players/human.rb'
+require_relative './players/player.rb'
+require_relative './ui/user_interface.rb'
+require_relative './ui/script.rb'
+
+def initializer(size, empty_char)
+  board = Board.new(size, empty_char)
+  checker = BoardChecker.new(size, empty_char)
```
Comment (zdavy) [inline lib/initializer.rb]: Seems strange to me that the board & the board checker take the same arguments to be initialized, is there any way to clean this up? Maybe passing a board to the methods that do the checking?

File: lib/initializer.rb [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  script = Script.new
+  user_interface = UserInterface.new(board, script)
+
+  user_interface.welcome
+  mode_selection = user_interface.what_mode?
+
+  unless(mode_selection == 'computerVScomputer')
+    player_selection = user_interface.who_goes_first?
+  end
+
```
Comment (zdavy) [inline lib/initializer.rb:25]: Though I did mention that setting up all the dependencies could be completed procedurally, I think that some of the "procedure" could be wrapped up into their own method. Something that just returns/yields some game options.

File: lib/initializer.rb [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+  behavior1, behavior2, other = setup_behavior(mode_selection, 'X', 'O', board, checker, user_interface)
+  player1, player2, you = setup_players(player_selection, behavior1, behavior2, user_interface)
```
Comment (zdavy) [inline lib/initializer.rb]: What do you think are the advantages/tradeoffs of multiple return values from a method have been?

File: lib/initializer.rb [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+  game = Game.new(board, checker, player1, player2, you, other)
```
Comment (zdavy) [inline lib/initializer.rb]: I think that the game initializer is great, it takes exactly what is needed for a game of tic-tac-toe, although the player1/player2/you/other feels a little redundant. Is there a way to merge the two concepts?

File: spec/initializer_spec.rb [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+describe 'initializer' do
+  before(:each) do
+    @board = Board.new(3, '-')
+    @checker = BoardChecker.new(3, '-')
+    @script = Script.new
+    @user_interface = UserInterface.new(@board, @script)
+    @ai_logic = AI.new(@checker, 'O', 'X')
+    @human_logic1 = Human.new(@board, @user_interface, 'X')
+    @human_logic2 = Human.new(@board, @user_interface, 'O')
+  end
```
Comment (zdavy) [inline spec/initializer_spec.rb]: These instance variables _work_ fine for the tests, but the rspec convention is to use `let` statements, they're lazy and aren't evaluated until they're utilized in the example.

File: lib/game.rb [hunk 7]
```diff
@@ -1,8 +1,9 @@
 class Game
-  def initialize(board, checker, player1, player2)
+  def initialize(board, checker, player1, player2, you, other)
```
Comment (zdavy) [inline lib/game.rb]: the you/other dichotomy is confusing to me, it feels like it's another representation of the player1/player2 concept, there should be a way to merge the ideas

File: lib/initializer.rb [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+  game.play
+end
+
+def setup_behavior(selection, token1, token2, board, checker, user_interface)
```
Comment (zdavy) [inline lib/initializer.rb]: The naming of the variables in this method (`is_human_player`,`is_ai_player`) makes it hard to follow logically what's happening in the method body.

File: lib/players/ai.rb [hunk 9]
```diff
@@ -7,7 +7,7 @@ def initialize(checker, token, min_token)
     @min_token = min_token
   end
 
-  def get_move(grid, turns_remaining)
+  def get_move(grid, turns_remaining, token)
```
Comment (zdavy) [inline lib/players/ai.rb]: Is this added token used anywhere?

Comment (zdavy) [reply lib/players/ai.rb]: Oh, I see, it's to adhere to the same interface as the human `get_move` method. When the provided arguments aren't necessary for both scenarios there should be another way to introduce that value to the instance that needs it (maybe through dependency injection?).

File: lib/players/player.rb [hunk 10]
```diff
@@ -13,10 +15,10 @@ def take_turn(board, turns_remaining)
   end
 
   def get_move(grid, turns_remaining)
-    move = @behavior.get_move(grid, turns_remaining)
+    move = @behavior.get_move(grid, turns_remaining, @token)
 
     if(@behavior.is_a? AI)
-      @user_interface.ai_move(move)
+      @user_interface.ai_move(move, @token)
     end
```
Comment (zdavy) [inline lib/players/player.rb]: Is there a way to solve this problem without type checking?

File: spec/user_interface_spec.rb [hunk 11]
```diff
@@ -22,6 +30,34 @@
...
+
+    it 'raises an error if selection is not valid' do
+      expect(user_interface.process_input(method(:callback), '4', script.game_modes)).to raise_error('Your selection is not valid.')
+    end
+  end
+
+  describe 'get_input' do
+    it 'puts message to STDOUT' do
+      allow(user_interface).to receive(:gets).and_return('1')
+      expect(STDOUT).to receive(:puts).with('message')
```
Comment (damonkelley) [inline spec/user_interface_spec.rb]: 🙀 

One way to help get rid of this mocking is to remove the hard dependency on `STDOUT` and `puts`. How could we do that?

An approach that I like is to eliminate hard dependencies on anything you don't own. In this case, that would `STDOUT` and `puts`.

File: lib/initializer.rb (hunk 2 above)
Comment (saralein) [inline lib/initializer.rb]: [Changed this to take the board](https://github.com/saralein/tic-tac-toe_ruby/pull/7/files#diff-6bfad8daa181b5e0a43f92245aa59e75R14)
